torchdata/stateful_dataloader/README.md

```diff
@@ -0,0 +1,215 @@
+    while self.i < self.limit:
+      val = int(torch.randint(high=self.high, size=(1,), generator=self.g))
+      self.i += 1
+      yield val
+
+  def load_state_dict(self, state_dict: Dict[str, Any]):
+    self.i = state_dict["i"]
+    self.g.set_state(state_dict["rng"])
+
+  def state_dict(self) -> Dict[str, Any]:
+    return {"i": self.i, "rng": self.g.get_state()}
+
+# Optional: save dataset random transform state
+class NoisyRange(torch.utils.data.Dataset):
+  def __init__(self, high: int, mean: float, std: float):
+    self.high, self.mean, self.std = high, torch.tensor([float(mean)]), float(std)
+
+  def __len__(self):
+    return self.high
+
+  def __getitem__(self, idx: int) -> float:
+    if not (0 <= idx < self.high):
+      raise IndexError()
+    x = torch.normal(self.mean, self.std)
+    noise = x.item()
+    return idx + noise
+
+  def load_state_dict(self, state_dict):
+    torch.set_rng_state(state_dict["rng"])
+
+  def state_dict(self):
+    return {"rng": torch.get_rng_state()}
+
+for num_workers in [0, 2]:
+  print(f"{num_workers=}")
+  dl = StatefulDataLoader(NoisyRange(5, 1, 1), sampler=MySampler(5, 1, 10),
+      batch_size=2, drop_last=False, num_workers=num_workers)
+
+  batches = []
+  for i, batch in enumerate(dl):
+    batches.append(batch)
+    if i == 2:
+      sd = dl.state_dict()
+
+  dl.load_state_dict(sd)
+  batches2 = list(dl)
+
+  print(batches[3:])
+  print(batches2)
+
+"""
+Output:
+num_workers=0
+[tensor([-0.4526,  3.7948], dtype=torch.float64), tensor([6.5494, 3.0470], dtype=torch.float64)]
+[tensor([-0.4526,  3.7948], dtype=torch.float64), tensor([6.5494, 3.0470], dtype=torch.float64)]
+num_workers=2
+[tensor([3.7412, 1.2438], dtype=torch.float64), tensor([4.4807, 4.0036], dtype=torch.float64)]
+[tensor([3.7412, 1.2438], dtype=torch.float64), tensor([4.4807, 4.0036], dtype=torch.float64)]
+"""
+```
+
+## Saving Custom State with Iterable-Style Datasets
+
+Tracking iteration order with Iterable-style datasets requires state from each worker-level instance of the dataset to
+be captured. You can define `state_dict/load_state_dict` methods on your dataset which capture worker-level state.
+`StatefulDataLoader` will handle aggregation across workers and distribution back to the workers. Calling
+`load_state_dict` requires `StatefulDataLoader` to have same `num_workers` as those of the provided `state_dict`.
+
+```py
+from typing import *
+import torch
+import torch.utils.data
+from torchdata.stateful_dataloader import StatefulDataLoader
+
+
+class MyIterableDataset(torch.utils.data.IterableDataset):
+  def __init__(self, high: int, seed: int):
+    self.high, self.seed = high, seed
+    self.g = torch.Generator()
+    self.i = 0
+
+  def __iter__(self):
+    worker_info = torch.utils.data.get_worker_info()
+    if worker_info is not None:
+      worker_id = worker_info.id
+      num_workers = worker_info.num_workers
+    else:
+      worker_id = 0
+      num_workers = 1
+    self.g.manual_seed(self.seed)
```

Comment:
Do we need to set the seed to be same for all workers? Can the one set by the worker loop in worker.py be used?

Reply:
Because of how this is set up (same array on all workers, iterate through based on worker_id) this is correct